app/window.py

```diff
@@ -362,17 +363,53 @@ def __init__(self, host):
   def drawLineNumbers(self):
     limit = min(self.rows,
         len(self.host.textBuffer.lines) - self.host.scrollRow)
-    color = app.color.get('line_number')
+    cursorBookmarkColorIndex = None
     for i in range(limit):
+      color = app.color.get('line_number')
+      rowBookmark = self.bookmarkForRow(self.host.scrollRow + i)
+      # Use a different color if the row is associated with a bookmark.
+      if rowBookmark:
+        color = app.color.get(rowBookmark[1].get('colorIndex'))
+        if self.host.cursorRow == self.host.scrollRow + i:
+          cursorBookmarkColorIndex = rowBookmark[1].get('colorIndex')
       self.addStr(i, 0, ' %5d ' % (self.host.scrollRow + i + 1), color)
     color = app.color.get('outside_document')
     for i in range(limit, self.rows):
       self.addStr(i, 0, '       ', color)
     cursorAt = self.host.cursorRow - self.host.scrollRow
     if 0 <= cursorAt < limit:
-      color = app.color.get('line_number_current')
+      if cursorBookmarkColorIndex:
+        if app.prefs.startup['numColors'] == 8:
+          color = app.color.get(cursorBookmarkColorIndex)
+        else:
+          color = app.color.get(cursorBookmarkColorIndex % 32 + 128)
+      else:
+        color = app.color.get('line_number_current')
       self.addStr(cursorAt, 1, '%5d' % (self.host.cursorRow + 1), color)
 
+  def bookmarkForRow(self, row):
+    """
+    Checks whether the specified row lies on a bookmark or not.
+
+    Args:
+      row (int): the row number you want to check between [0, numLines - 1].
+
+    Returns:
+      The bookmark that the row is associated with or None if there isn't one.
+    """
```

Comment:
Since bookmarks tend to come in blocks of rows, would it be more efficient to have a function that returned a list of bookmarked rows on the current screen?

i.e. 
- args are beginRow, endRow
- use bisect to find beginRow
- then 'walk' the bookmarks until endRow
- return list of bookmarked rows

Reply:
I ended up returning a list of bookmarks itself, since we need the actual bookmarks when rendering the line numbers. I think it should perform a bit better now.